### ISSUE TITLE
perlsource.pod: Misleading discussion of what should go into ext/

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13475]
Migrated from [rt.perl.org#120808](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120808) (status was 'resolved')

Searchable as RT120808$

**Comment by p5pRT** [comment]
### From @jkeenan

I posed this question on perl\.perl5porters earlier this year and got no 
responses \-\- except one from rjbs asking me if I had received any 
responses and suggesting an RT if I had not\.  Hence\, \.\.\.

pod/perlsource\.pod describes the 'ext/' directory in the Perl 5 
distribution thus&#8203;:

\#\#\#\#\#
=item \* F\<ext/>

This directory contains XS\-using modules which are only released as
part of the core\. These modules generally have their F\<Makefile\.PL> and
are laid out more like a typical CPAN module\.
\#\#\#\#\#

This is misleading because \(as of my last count in May 2013\)\, there are 
35 subdirectories underneath ext/ of which only 28 have XS\.  The 7 
subdirs which do \*not\* have XS are&#8203;:

Errno
FileCache
IPC\-Open3
Pod\-Functions
Pod\-Html
Tie\-Memoize
Win32CORE

My sense is that\, until relatively recent times\, 'ext/' always held 
modules with XS\.  That's certainly the simplest interpretation of the 
first sentence in the paragraph from perlsource\.pod quoted above\.

Note that in the following paragraph in perlsource\.pod\, we have this&#8203;:

\#\#\#\#\#
=item \* F\<dist/>

This directory is for dual\-life modules where the blead source is
canonical\. Note that some modules in this directory may not yet have
been released separately on CPAN\.
\#\#\#\#\#

It seems to me that the 7 non\-XS using libraries under ext/ could just 
as easily be described as "not yet \[having\] been released separately on 
CPAN" \-\- which would mean that they would fit better under dist/ than ext/\.

If we can get a decision on what the layout should be\, I will either&#8203;: 
\(a\) revise the description of 'ext/' to de\-emphasize the XS\-using part 
of it; or \(b\) set about moving the 7 non\-XS using modules to 'dist/'\.

Choose\!

Thank you very much\.
Jim Keenan

Now\, I could certainly live with a description of 'ext/' as holding 
libraries with a CPAN\-style directory and file layout not maintained or 
distributed on CPAN\.  That wouldn't be a very elegant description\, and 
it would sever the tie between 'ext' and 'eXternal Subroutine'\.  But it 
would be more accurate than what we have now\.  And this wouldn't require 
us to move any libraries from one of the subdirs to another\.

**Comment by p5pRT** [comment]
### From @jkeenan

$ perl \-V
Summary of my perl5 \(revision 5 version 18 subversion 0\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=8\.11\.0\, archname=darwin\-2level
&nbsp;    uname='darwin macintosh\-9\.local 8\.11\.0 darwin kernel version 8\.11\.0&#8203;: wed oct 10 18&#8203;:26&#8203;:00 pdt 2007; root&#8203;:xnu\-792\.24\.17~1release\_ppc power macintosh powerpc '
&nbsp;    config\_args='\-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-I/opt/local/include'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-I/opt/local/include'
&nbsp;    ccversion=''\, gccversion='4\.0\.1 \(Apple Computer\, Inc\. build 5250\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-L/usr/local/lib \-L/opt/local/lib'
&nbsp;    libpth=/usr/local/lib /opt/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lc
&nbsp;    perllibs=\-ldl \-lm \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-L/opt/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_PRESERVE\_IVUV PERL\_SAWAMPERSAND USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_PERLIO USE\_PERL\_ATOF
&nbsp;  Built under darwin
&nbsp;  Compiled at May 18 2013 12&#8203;:31&#8203;:26
&nbsp;  %ENV&#8203;:
&nbsp;    PERLBREW\_BASHRC\_VERSION="0\.59"
&nbsp;    PERLBREW\_HOME="/Users/jimk/\.perlbrew"
&nbsp;    PERLBREW\_MANPATH=""
&nbsp;    PERLBREW\_PATH="/Users/jimk/perl5/perlbrew/bin"
&nbsp;    PERLBREW\_ROOT="/Users/jimk/perl5/perlbrew"
&nbsp;    PERLBREW\_VERSION="0\.59"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.18\.0/darwin\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.18\.0
&nbsp;    /usr/local/lib/perl5/5\.18\.0/darwin\-2level
&nbsp;    /usr/local/lib/perl5/5\.18\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.16\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.14\.2
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.14\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.10\.1
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.10\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.6
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Dec 16 19&#8203;:24&#8203;:18 2013\, jkeen@&#8203;verizon\.net wrote&#8203;:
> It seems to me that the 7 non\-XS using libraries under ext/ could just 
> as easily be described as "not yet \[having\] been released separately on 
> CPAN" \-\- which would mean that they would fit better under dist/ than ext/\.

At least to me\, the main distinction between dist and ext is that code under dist needs to work with older versions\, whereas we can do whatever we like \(use new shiny features\) with ext\.  Maybe that could be incorporated into a new description for ext\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

On Mon Dec 16 19&#8203;:24&#8203;:18 2013\, jkeen@&#8203;verizon\.net wrote&#8203;:
> I posed this question on perl\.perl5porters earlier this year and got no 
> responses \-\- except one from rjbs asking me if I had received any 
> responses and suggesting an RT if I had not\.  Hence\, \.\.\.
> 
> pod/perlsource\.pod describes the 'ext/' directory in the Perl 5 
> distribution thus&#8203;:
> 
> \#\#\#\#\#
> =item \* F\<ext/>
> 
> This directory contains XS\-using modules which are only released as
> part of the core\. These modules generally have their F\<Makefile\.PL> and
> are laid out more like a typical CPAN module\.
> \#\#\#\#\#
> 
> This is misleading because \(as of my last count in May 2013\)\, there are 
> 35 subdirectories underneath ext/ of which only 28 have XS\.  The 7 
> subdirs which do \*not\* have XS are&#8203;:
> 
> Errno
> FileCache
> IPC\-Open3
> Pod\-Functions
> Pod\-Html
> Tie\-Memoize
> Win32CORE
> 
> My sense is that\, until relatively recent times\, 'ext/' always held 
> modules with XS\.  That's certainly the simplest interpretation of the 
> first sentence in the paragraph from perlsource\.pod quoted above\.
> 
> Note that in the following paragraph in perlsource\.pod\, we have this&#8203;:
> 
> \#\#\#\#\#
> =item \* F\<dist/>
> 
> This directory is for dual\-life modules where the blead source is
> canonical\. Note that some modules in this directory may not yet have
> been released separately on CPAN\.
> \#\#\#\#\#
> 
> It seems to me that the 7 non\-XS using libraries under ext/ could just 
> as easily be described as "not yet \[having\] been released separately on 
> CPAN" \-\- which would mean that they would fit better under dist/ than ext/\.
> 
> If we can get a decision on what the layout should be\, I will either&#8203;: 
> \(a\) revise the description of 'ext/' to de\-emphasize the XS\-using part 
> of it; or \(b\) set about moving the 7 non\-XS using modules to 'dist/'\.
> 
> Choose\!
> 
> Thank you very much\.
> Jim Keenan
> 
> Now\, I could certainly live with a description of 'ext/' as holding 
> libraries with a CPAN\-style directory and file layout not maintained or 
> distributed on CPAN\.  That wouldn't be a very elegant description\, and 
> it would sever the tie between 'ext' and 'eXternal Subroutine'\.  But it 
> would be more accurate than what we have now\.  And this wouldn't require 
> us to move any libraries from one of the subdirs to another\.
> 

Win32CORE has a \.c files with xsubs\. So it basically has XS\.

Pod\-Html should be on CPAN or dual lifed\. Pod\-Html\, which is used by ActivePerl for their pretty \(CSSed\) HTML docs\, suffered some bug fixes and some regressions between 5\.14 and 5\.16\. I'm not sure where to even file a ticket\, or if it can ever be fixed since Pod\-Html is not distributed on CPAN and therefore isn't upgradable\.
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Dec 16 19&#8203;:24&#8203;:18 2013\, jkeen@&#8203;verizon\.net wrote&#8203;:
> Now\, I could certainly live with a description of 'ext/' as holding 
> libraries with a CPAN\-style directory and file layout not maintained or 
> distributed on CPAN\.  That wouldn't be a very elegant description\, and 
> it would sever the tie between 'ext' and 'eXternal Subroutine'\.  But it 
> would be more accurate than what we have now\.  And this wouldn't require 
> us to move any libraries from one of the subdirs to another\.

I believe this is the correct way to handle it\.

On Mon Dec 16 22&#8203;:53&#8203;:48 2013\, bulk88 wrote&#8203;:
> Pod\-Html should be on CPAN or dual lifed\. Pod\-Html\, which is used by
> ActivePerl for their pretty \(CSSed\) HTML docs\, suffered some bug fixes
> and some regressions between 5\.14 and 5\.16\. I'm not sure where to even
> file a ticket\, or if it can ever be fixed since Pod\-Html is not
> distributed on CPAN and therefore isn't upgradable\.

I don't think anyone would object to you packaging Pod\-Html for CPAN\.

Tony

**Comment by p5pRT** [comment]
### From @rjbs

\* Tony Cook via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-12\-17T17&#8203;:57&#8203;:53\]
> On Mon Dec 16 22&#8203;:53&#8203;:48 2013\, bulk88 wrote&#8203;:
> > Pod\-Html should be on CPAN or dual lifed\. Pod\-Html\, which is used by
> > ActivePerl for their pretty \(CSSed\) HTML docs\, suffered some bug fixes
> > and some regressions between 5\.14 and 5\.16\. I'm not sure where to even
> > file a ticket\, or if it can ever be fixed since Pod\-Html is not
> > distributed on CPAN and therefore isn't upgradable\.
> 
> I don't think anyone would object to you packaging Pod\-Html for CPAN\.

I would owe the packager a beer\, juice\, or other beverage or snack\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @jkeenan

On 12/19/13 9&#8203;:34 PM\, Ricardo Signes wrote&#8203;:
> \* Tony Cook via RT\<perlbug\-followup@&#8203;perl\.org>  \[2013\-12\-17T17&#8203;:57&#8203;:53\]
>> On Mon Dec 16 22&#8203;:53&#8203;:48 2013\, bulk88 wrote&#8203;:
>>> Pod\-Html should be on CPAN or dual lifed\. Pod\-Html\, which is used by
>>> ActivePerl for their pretty \(CSSed\) HTML docs\, suffered some bug fixes
>>> and some regressions between 5\.14 and 5\.16\. I'm not sure where to even
>>> file a ticket\, or if it can ever be fixed since Pod\-Html is not
>>> distributed on CPAN and therefore isn't upgradable\.
>>
>> I don't think anyone would object to you packaging Pod\-Html for CPAN\.
>
> I would owe the packager a beer\, juice\, or other beverage or snack\.
>

The discussion has gone a bit off topic\.  I'd like to discuss Pod\-Html\, 
but in a separate thread\.   So\, too get back on topic\, let me ask&#8203;:

rjbs&#8203;:  Do you have feedback on my OP on pod/perlsource\.pod and on Father 
C's and Tony's comments?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @rjbs

On Fri Dec 20 12&#8203;:55&#8203;:22 2013\, jkeen@&#8203;verizon\.net wrote&#8203;:
> rjbs&#8203;:  Do you have feedback on my OP on pod/perlsource\.pod and on Father 
> C's and Tony's comments?

Yes&#8203;: I agree with your suggestion and with FC's elaboration\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Feb 10 14&#8203;:47&#8203;:51 2014\, rjbs wrote&#8203;:
> On Fri Dec 20 12&#8203;:55&#8203;:22 2013\, jkeen@&#8203;verizon\.net wrote&#8203;:
> > rjbs&#8203;:  Do you have feedback on my OP on pod/perlsource\.pod and on Father 
> > C's and Tony's comments?
> 
> Yes&#8203;: I agree with your suggestion and with FC's elaboration\.

Okay\, then let's get concrete\.  Please evaluate the patch attached\.  \(I'll apply it to blead in 7 days unless there is an objection\.\)

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>120808-0001-Clarify-distinction-between-contents-of-dist-and-ext.patch</summary>

```diff
From 549fcc75f5665e162dd09e8d3eb5e5a66f69f428 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Tue, 11 Feb 2014 01:50:47 +0100
Subject: [PATCH] Clarify distinction between contents of dist/ and ext/.

For: RT #120808
---
 pod/perlsource.pod |   21 +++++++++++++--------
 1 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/pod/perlsource.pod b/pod/perlsource.pod
index 81e3e94..aa351dd 100644
--- a/pod/perlsource.pod
+++ b/pod/perlsource.pod
@@ -46,15 +46,20 @@ their tests, unlike other core modules.
 
 =item * F<ext/>
 
-This directory contains XS-using modules which are only released as
-part of the core. These modules generally have their F<Makefile.PL> and
-are laid out more like a typical CPAN module.
+Like F<lib/>, this directory contains modules which are only released
+as part of the core.  Unlike F<lib/>, however, a module under F<ext/>
+generally has a CPAN-style directory- and file-layout and its own
+F<Makefile.PL>.  There is no expectation that a module under F<ext/>
+will work with earlier versions of Perl 5.  Hence, such a module may
+take full advantage of syntactical and other improvements in Perl 5
+blead.
 
 =item * F<dist/>
 
 This directory is for dual-life modules where the blead source is
 canonical. Note that some modules in this directory may not yet have
-been released separately on CPAN.
+been released separately on CPAN.  Modules under F<dist/> are expected
+to work with earlier versions of Perl 5.
 
 =item * F<cpan/>
 
@@ -118,10 +123,10 @@ other directories.
 
 =item * F<t/opbasic/>
 
-Tests for perl's built in functions which, like those in F<t/op/>, do not fit
-into any of the other directories, but which, in addition, cannot use
-F<t/test.pl>,as that program depends on functionality which the
-test file itself is testing.
+Tests for perl's built in functions which, like those in F<t/op/>, do
+not fit into any of the other directories, but which, in addition,
+cannot use F<t/test.pl>,as that program depends on functionality which
+the test file itself is testing.
 
 =item * F<t/re/>
 
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

On Mon Feb 10 18&#8203;:29&#8203;:49 2014\, jkeenan wrote&#8203;:
> Okay\, then let's get concrete\.  Please evaluate the patch attached\.
> \(I'll apply it to blead in 7 days unless there is an objection\.\)

Looks fine to me\.

If I were to pick a nit\, it would be this&#8203;:

\+been released separately on CPAN\.  Modules under F\<dist/> are expected
\+to work with earlier versions of Perl 5\.

I might say "will often work" or "should make an effort to work\."  The language "are expected" makes me wonder "expected by whom?" and whether it implies any obligation\.  I don't think there is a blanket obligation to support earlier perls\, but that it we should not gratuitously cease to support them\.

At any rate\, the patch is a marked improvement on the accuracy of the documentation\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @karenetheridge

On Mon\, Feb 10\, 2014 at 06&#8203;:29&#8203;:49PM \-0800\, James E Keenan via RT wrote&#8203;:
> Okay\, then let's get concrete\.  Please evaluate the patch attached\.  \(I'll apply it to blead in 7 days unless there is an objection\.\)

With these changes\, will it be possible to distinguish \(in\, say\,
Module&#8203;::CoreList\) whether a module is dual\-lifed or not by looking at which
directory it lives in?

If yes\, it would also be useful to enforce the directory distinctions by
testing if particular modules are indexed by PAUSE\, and whether the
associated distribution is its own dual\-lifed distribution or the latest
stable perl release\. I would be happy to write a patch if someone can
suggest where such a test should live and when it would run\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Feb 16 13&#8203;:28&#8203;:10 2014\, perl@&#8203;froods\.org wrote&#8203;:
> On Mon\, Feb 10\, 2014 at 06&#8203;:29&#8203;:49PM \-0800\, James E Keenan via RT wrote&#8203;:
> > Okay\, then let's get concrete\.  Please evaluate the patch attached\.
> > \(I'll apply it to blead in 7 days unless there is an objection\.\)
> 
> With these changes\, will it be possible to distinguish \(in\, say\,
> Module&#8203;::CoreList\) whether a module is dual\-lifed or not by looking at
> which
> directory it lives in?
> 

No\.

But that's no change from the current situation\, where dual\-lifed modules may exist in either cpan/ or dist/\.

My patch is just a documentation patch and does not move any modules from one tree to another\.  The objective is simply to describe the contents of the various trees more accurately than pod/perlsource\.pod does now\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Feb 14 18&#8203;:06&#8203;:56 2014\, rjbs wrote&#8203;:
> On Mon Feb 10 18&#8203;:29&#8203;:49 2014\, jkeenan wrote&#8203;:
> > Okay\, then let's get concrete\.  Please evaluate the patch attached\.
> > \(I'll apply it to blead in 7 days unless there is an objection\.\)
> 
> Looks fine to me\.
> 
> If I were to pick a nit\, it would be this&#8203;:
> 
> \+been released separately on CPAN\.  Modules under F\<dist/> are
> expected
> \+to work with earlier versions of Perl 5\.
> 
> I might say "will often work" or "should make an effort to work\."  The
> language "are expected" makes me wonder "expected by whom?" and
> whether it implies any obligation\.  I don't think there is a blanket
> obligation to support earlier perls\, but that it we should not
> gratuitously cease to support them\.
> 
> At any rate\, the patch is a marked improvement on the accuracy of the
> documentation\.

Patch\, incorporating rjbs's suggestion\, was applied to blead in commit 259799cf6d5d97ce56449b0c6cbe16e4fd845fb3\.

Closing ticket\.  Any other issues concerning the directory structure should be raised on p5p list or in a new RT\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 16 February 2014 21&#8203;:27\, Karen Etheridge \<perl@&#8203;froods\.org> wrote&#8203;:
> On Mon\, Feb 10\, 2014 at 06&#8203;:29&#8203;:49PM \-0800\, James E Keenan via RT wrote&#8203;:
>> Okay\, then let's get concrete\.  Please evaluate the patch attached\.  \(I'll apply it to blead in 7 days unless there is an objection\.\)
>
> With these changes\, will it be possible to distinguish \(in\, say\,
> Module&#8203;::CoreList\) whether a module is dual\-lifed or not by looking at which
> directory it lives in?
>
> If yes\, it would also be useful to enforce the directory distinctions by
> testing if particular modules are indexed by PAUSE\, and whether the
> associated distribution is its own dual\-lifed distribution or the latest
> stable perl release\. I would be happy to write a patch if someone can
> suggest where such a test should live and when it would run\.

My understanding was that a module is dual\-lived if and only if it has
a DISTRIBUTION field in Porting/Maintainers\.pl\. \(I have no idea
whether that information is available via Module&#8203;::CoreList\, though\.\)
